man/meshd.8.pod

  Copyright 2011-2015 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.


=head1 NAME

meshd - Clockwork Mesh Master

=head1 SYNOPSIS

B<meshd> [options]

B<meshd> -c /path/to/meshd.conf

B<meshd> -Fvvv

=head1 DESCRIPTION

#INTRO

B<meshd> is the Clockwork Mesh Master server.  It handles COMMAND broadcast
and query submission / result retrieval for Mesh clients.

=head1 OPTIONS

=over

=item B<-V>, B<--version>

Print version and copyright information.

=item B<-c> B<--config> I</path/to/meshd.conf>

Specify the path to an alternate configuration file.  If not
given, the default configuration file (/etc/clockwork/meshd.conf)
is used instead.

=item B<-q>, B<--quiet>

Disable logging.

=item B<-v>, B<--verbose>

Increase verbosity by one level.  Can be used more than once.

=item B<-t>, B<--test>

Test meshd configuration and environment.  This checks both configuration
file syntax and the validity of the specified B<security.cert>.

=item B<-S>, B<--show-config>

Once B<meshd> has determined its running configuration, have
it print that to standard output and exit.

=item B<-F>, B<--foreground>

Normally, B<meshd> forks into the background and detaches from
the controlling terminal, so that it can run as a system service.
This flag disabled that behavior.  As a side effect, all logging
will be forced to go to standard output, regardless of what syslog
facility is specified in the configuration file.

=back

=head1 FILES

=over

=item B</etc/clockwork/meshd.conf>

Default configuration file that B<meshd> reads on start up.
See B<meshd.conf>(5) for details.

=back

=head1 AUTHOR

#AUTHOR

=cut


CHANGELOG

origin/master                                              runtime 20150202

  (none)



3.0.10       2014-02-04                                    runtime 20150202

  [ENHANCEMENTS]

  - New Mesh commands for managing packages and services

    `package <name> <action> [version]` lets remote administrators manage
    and query the installed system packages.  Additionally, `package
    recache` will rebuild the package cache (i.e. apt-get update, or yum
    clean all).

    `service <name> <action>` deals with service run/boot status.



3.0.9        2014-02-02                                    runtime 20150202

  [ENHANCEMENTS]

  - New `sha1` opcode
    Useful for generating checksums of in-memory strings (like file names)

  - Modify `property` opcode to return pragma values
    If a property with the desired name isn't found, we should return the
    named pragma's value.  This lets us use pragmas as globals vars (without
    scope).  It also lets us move more functionality into stdlib.pn

  - Offline/Cache Mode
    cogd now supports cache/offline mode.  If a previous CFM run was
    successfull, the Pendulum bytecode image that represents the policy will
    be cached.  Subsequent runs that fail to connect to a clockd master can
    then fallback on this cached policy for local enforcement of the "last
    known good state"

  - Local File Caching support
    If files are marked 'cache: yes' in their resource definitions, cogd
    will keep the desired contents in a local cache directory, so that a
    future run can still enforce system state even when disconnected from a
    clockd policy master daemon.



3.0.8        2014-02-01                                    runtime 20150201

  [ENHANCMENTS]

  - New `augeas.exists?` opcode
    Like the 20150119 opcode `augeas.find`, except that it has no output
    parameter and therefore does no heap-allocation.

  - Introduce proper JSON and XML/CDATA escaping in mesh output routines



3.0.7        2014-02-01                                    runtime 20150131

  [ENHANCEMENTS]

  - Set a umask in cogd
    cogd now sets a more sane umask, which can be specified via a new
    configuration directive, 'umask'.  Defaults to 0022.

  - pn --output flag
    The pn utility, in assembler (-S) mode, now supports -o / --output
    for specifying where to write the pendulum bytecode image.  This allows
    callers to short-circuit the auto-naming ("just add .S") logic, and was
    needed to support VPATH builds in automake.

  - New clockwork.* facts
    The new clockwork.version, clockwork.runtime and clockwork.protocol
    facts can be used to query for version (x.y.z), runtime (YYYYMMDD) and
    network protocol version (N) in play.

  - New `runtime` opcode for determining version of Pendulum runtime
    The runtime version will be set to date of finalization of new
    instruction set prior to release.  This means the each runtime version
    is a fixed specification of interface (but not necessarily behavior),
    and pendulum code can use this new opcode to determine if the runtime is
    new enough before using new opcodes.

  - New `fs.mkparent` opcode
    Intelligently creates missing parent directories of its first argument,
    using the ownership and mode of the closest existing ancestor directory.
    This emulates pre-3.0 Clockwork behavior that was missed during the
    conversion to the register machine implementation.

  - Better error / output handling from `localsys` in stdlib.pn

  [BUG FIXES]

  - Memory leaks plugged
    cogd no longer leaks the bytecode image it received from clockd (big).
    SHA1.FAIL PDUs are no longer leaked by `remote.sha1` opcode.
    SHA1 checksums are no longer leaked by `remote.sha1` opcode.

  - Fix some vim pnasm highlighting issues

  - Remove superfluous debugging from stdlib.pn



3.0.6        2014-01-30                                    runtime 20150119

  [BUG FIXES]

  - Fix misfiring notifies for dependency forward tracking.
    Now, if a service depends on a config file, changing the config file
    will cause cogd to properly reload/restart the dependent service.

  - Properly set "changed" (lastchg) attribute on new users
    Without it, new users are required to change their passwords immediately
    upon login.



3.0.5        2014-01-29                                    runtime 20150119

  [BUG FIXES]

  - Fix an issue in stdlib.pn's res.file.contents function, whereby file
    resources that come with verification steps (as /etc/sudoers often do)
    don't transfer mode or ownership (UID/GID) to the temporary file before
    renaming it over the top of the original file.  This causes havoc on
    files like /etc/sudoers, which *must* have specific permissions.

  - Fix typo (retrieve not 'retreive') in diagnostic message for remote.file.



3.0.4        2014-01-28                                    runtime 20150119

  [BUG FIXES]

  - cogd now honors `timeout` when waiting for responses from clockd
    masters, during remediation.  This fixes several stall-out conditions
    where the clockd master goes away mid-run, and the cogd process hangs
    waiting for a response that will never come.



3.0.3        2014-01-27                                    runtime 20150119

  [ENHANCEMENTS]

  - SIGHUP-to-reload Support for cogd
    cogd supports use of `kill -HUP $pid` to cause it to reload
    configuration while running.

  - SIGHUP-to-reload Support for clockd
    clockd supports use of `kill -HUP $pid` to cause it to reload
    configuration and parse a new policy manifest while running.
    Intelligently handles client connections for in-flight remediation.



3.0.2        2014-01-22                                    runtime 20150119

  [ENHANCEMENTS]

  - Support '-%>' for newline suppression in ERB templates
    Any ERB ruby snippet ending with '-%>' (instead of just '%>'), will
    ignore trailing whitespace.

  [BUG FIXES]


src/meshd.c
/*
  Copyright 2011-2015 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "clockwork.h"
#include "vm.h"
#include "mesh.h"
#include <getopt.h>

#define DEFAULT_CONFIG_FILE "/etc/clockwork/meshd.conf"

#define MODE_RUN  0
#define MODE_DUMP 1
#define MODE_TEST 2

static inline mesh_server_t *s_server_new(int argc, char **argv)
{
	char *t;
	int mode = MODE_RUN, daemon = 1;

	LIST(config);
	config_set(&config, "control",             "*:2315");
	config_set(&config, "broadcast",           "*:2316");
	config_set(&config, "client.connections",  "2048");
	config_set(&config, "client.expiration",   "600");
	config_set(&config, "pendulum.inc",        PENDULUM_INCLUDE);
	config_set(&config, "syslog.ident",        "meshd");
	config_set(&config, "syslog.facility",     "daemon");
	config_set(&config, "syslog.level",        "error");
	config_set(&config, "auth.service",        "clockwork");
	config_set(&config, "auth.trusted",        "/etc/clockwork/auth/trusted");
	config_set(&config, "security.cert",       "/etc/clockwork/certs/meshd");
	config_set(&config, "acl.global",          "/etc/clockwork/global.acl");
	config_set(&config, "pidfile",             "/var/run/meshd.pid");

	log_open(config_get(&config, "syslog.ident"), "stderr");
	log_level(0, (getenv("MESHD_DEBUG") ? "debug" : "error"));
	logger(LOG_DEBUG, "default configuration:");
	logger(LOG_DEBUG, "  broadcast           %s", config_get(&config, "broadcast"));
	logger(LOG_DEBUG, "  control             %s", config_get(&config, "control"));
	logger(LOG_DEBUG, "  client.connections  %s", config_get(&config, "client.connections"));
	logger(LOG_DEBUG, "  client.expiration   %s", config_get(&config, "client.expiration"));
	logger(LOG_DEBUG, "  pendulum.inc        %s", config_get(&config, "pendulum.inc"));
	logger(LOG_DEBUG, "  syslog.ident        %s", config_get(&config, "syslog.ident"));
	logger(LOG_DEBUG, "  syslog.facility     %s", config_get(&config, "syslog.facility"));
	logger(LOG_DEBUG, "  syslog.level        %s", config_get(&config, "syslog.level"));
	logger(LOG_DEBUG, "  auth.service        %s", config_get(&config, "auth.service"));
	logger(LOG_DEBUG, "  auth.trusted        %s", config_get(&config, "auth.trusted"));
	logger(LOG_DEBUG, "  security.cert       %s", config_get(&config, "security.cert"));
	logger(LOG_DEBUG, "  acl.global          %s", config_get(&config, "acl.global"));
	logger(LOG_DEBUG, "  pidfile             %s", config_get(&config, "pidfile"));


	logger(LOG_DEBUG, "processing command-line options");
	const char *short_opts = "?hvqVFSt" "c:";
	struct option long_opts[] = {
		{ "help",        no_argument,       NULL, 'h' },
		{ "verbose",     no_argument,       NULL, 'v' },
		{ "quiet",       no_argument,       NULL, 'q' },
		{ "version",     no_argument,       NULL, 'V' },
		{ "foreground",  no_argument,       NULL, 'F' },
		{ "show-config", no_argument,       NULL, 'S' },
		{ "test",        no_argument,       NULL, 't' },
		{ "config",      required_argument, NULL, 'c' },
		{ 0, 0, 0, 0 },
	};
	int verbose = -1;
	const char *config_file = DEFAULT_CONFIG_FILE;
	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			logger(LOG_DEBUG, "handling -h/-?/--help");
			printf("meshd, part of clockwork v%s runtime %i\n",
				CLOCKWORK_VERSION, CLOCKWORK_RUNTIME);
			printf("Usage: meshd [-?hvqVFSt] [-c filename]\n\n");
			printf("Options:\n");
			printf("  -?, -h               show this help screen\n");
			printf("  -V, --version        show version information and exit\n");
			printf("  -v, --verbose        increase logging verbosity\n");
			printf("  -q, --quiet          disable logging\n");
			printf("  -F, --foreground     don't daemonize, run in the foreground\n");
			printf("  -S, --show-config    print configuration and exit\n");
			printf("  -t, --test           test configuration\n");
			printf("  -c filename          set configuration file (default: " DEFAULT_CONFIG_FILE ")\n");
			exit(0);

		case 'v':
			if (verbose < 0) verbose = 0;
			verbose++;
			logger(LOG_DEBUG, "handling -v/--verbose (modifier = %i)", verbose);
			break;

		case 'q':
			verbose = 0;
			logger(LOG_DEBUG, "handling -q/--quiet (modifier = %i)", verbose);
			break;

		case 'V':
			logger(LOG_DEBUG, "handling -V/--version");
			printf("meshd (Clockwork) v%s runtime %i\n"
			       "Copyright (C) 2014 James Hunt\n",
			       CLOCKWORK_VERSION, CLOCKWORK_RUNTIME);
			exit(0);
			break;

		case 'c':
			logger(LOG_DEBUG, "handling -c/--config; replacing '%s' with '%s'",
				config_file, optarg);
			config_file = optarg;
			break;

		case 'F':
			logger(LOG_DEBUG, "handling -F/--foreground; turning off daemonize behavior");
			daemon = 0;
			break;

		case 'S':
			logger(LOG_DEBUG, "handling -S/--show-config");
			mode = MODE_DUMP;
			break;

		case 't':
			logger(LOG_DEBUG, "handling -t/--test");
			mode = MODE_TEST;
			break;
		}
	}
	logger(LOG_DEBUG, "option processing complete");


	logger(LOG_DEBUG, "parsing meshd configuration file '%s'", config_file);
	FILE *io = fopen(config_file, "r");
	if (!io) {
		logger(LOG_WARNING, "Failed to read configuration from %s: %s",
			config_file, strerror(errno));
		logger(LOG_WARNING, "Using default configuration");

	} else {
		if (config_read(&config, io) != 0) {
			logger(LOG_ERR, "Unable to parse %s");
			exit(1);
		}
		fclose(io);
		io = NULL;
	}


	logger(LOG_DEBUG, "determining adjusted log level/facility");
	if (verbose < 0) verbose = 0;
	t = config_get(&config, "syslog.level");
	logger(LOG_DEBUG, "configured log level is '%s', verbose modifier is %+i", t, verbose);
	int level = log_level_number(t);
	if (level < 0) {
		logger(LOG_WARNING, "'%s' is not a recognized log level, falling back to 'error'", t);
		level = LOG_ERR;
	}
	level += verbose;
	logger(LOG_DEBUG, "adjusted log level is %s (%i)",
		log_level_name(level), level);
	if (!daemon) {
		logger(LOG_DEBUG, "Running in --foreground mode; forcing all logging to stdout");
		config_set(&config, "syslog.facility", "stdout");
	}
	if (mode == MODE_DUMP) {
		logger(LOG_DEBUG, "Running in --show-config mode; forcing all logging to stderr");
		config_set(&config, "syslog.facility", "stderr");
	}
	if (mode == MODE_TEST) {
		logger(LOG_DEBUG, "Running in --test mode; forcing all logging to stderr");
		config_set(&config, "syslog.facility", "stderr");
	}
	logger(LOG_DEBUG, "redirecting to %s log as %s",
		config_get(&config, "syslog.facility"),
		config_get(&config, "syslog.ident"));

	log_open(config_get(&config, "syslog.ident"),
	            config_get(&config, "syslog.facility"));
	log_level(level, NULL);


	logger(LOG_INFO, "meshd starting up");


	if (mode == MODE_DUMP) {
		printf("control             %s\n", config_get(&config, "control"));
		printf("broadcast           %s\n", config_get(&config, "broadcast"));
		printf("client.connections  %s\n", config_get(&config, "client.connections"));
		printf("client.expiration   %s\n", config_get(&config, "client.expiration"));
		printf("pendulum.inc        %s\n", config_get(&config, "pendulum.inc"));
		printf("syslog.ident        %s\n", config_get(&config, "syslog.ident"));
		printf("syslog.facility     %s\n", config_get(&config, "syslog.facility"));
		printf("syslog.level        %s\n", config_get(&config, "syslog.level"));
		printf("auth.service        %s\n", config_get(&config, "auth.service"));
		printf("auth.trusted        %s\n", config_get(&config, "auth.trusted"));
		printf("security.cert       %s\n", config_get(&config, "security.cert"));
		printf("acl.global          %s\n", config_get(&config, "acl.global"));
		printf("pidfile             %s\n", config_get(&config, "pidfile"));
	}


	cert_t *cert = cert_read(config_get(&config, "security.cert"));
	if (!cert) {
		logger(LOG_ERR, "%s: %s", config_get(&config, "security.cert"),
			errno == EINVAL ? "Invalid Clockwork certificate" : strerror(errno));
		exit(1);
	}
	if (!cert->seckey) {
		logger(LOG_ERR, "%s: No secret key found in certificate",
			config_get(&config, "security.cert"));
		exit(1);
	}
	if (!cert->ident) {
		logger(LOG_ERR, "%s: No identity found in certificate",
			config_get(&config, "security.cert"));
		exit(1);
	}

	if (mode == MODE_TEST) {
		printf("Configuration OK\n");
		exit(0);
	}


	if (daemon)
		daemonize(config_get(&config, "pidfile"), "root", "root");

	mesh_server_t *s = mesh_server_new(NULL);
	int rc = mesh_server_setopt(s, MESH_SERVER_CERTIFICATE, cert, sizeof(cert_t));
	assert(rc == 0);
	cert_free(cert);

	size_t len = atoi(config_get(&config, "client.connections"));
	rc = mesh_server_setopt(s, MESH_SERVER_CACHE_SIZE, &len, sizeof(len));
	assert(rc == 0);

	int32_t life = atoi(config_get(&config, "client.expiration"));
	rc = mesh_server_setopt(s, MESH_SERVER_CACHE_LIFE, &life, sizeof(life));
	assert(rc == 0);

	t = config_get(&config, "pendulum.inc");
	rc = mesh_server_setopt(s, MESH_SERVER_PN_INCLUDE, t, strlen(t));
	assert(rc == 0);

	t = config_get(&config, "auth.service");
	rc = mesh_server_setopt(s, MESH_SERVER_PAM_SERVICE, t, strlen(t));
	assert(rc == 0);

	t = config_get(&config, "auth.trusted");
	rc = mesh_server_setopt(s, MESH_SERVER_TRUSTDB, t, strlen(t));
	assert(rc == 0);

	t = config_get(&config, "acl.global");
	rc = mesh_server_setopt(s, MESH_SERVER_GLOBAL_ACL, t, strlen(t));
	assert(rc == 0);

	t = string("tcp://%s", config_get(&config, "control"));
	logger(LOG_DEBUG, "binding control channel %s", t);
	rc = mesh_server_bind_control(s, t);
	assert(rc == 0);
	free(t);

	t = string("tcp://%s", config_get(&config, "broadcast"));
	logger(LOG_DEBUG, "binding to broadcast channel %s", t);
	rc = mesh_server_bind_broadcast(s, t);
	assert(rc == 0);
	free(t);

	config_done(&config);
	logger(LOG_INFO, "meshd running");
	return s;
}

int main(int argc, char **argv)
{
#ifdef UNIT_TESTS
	/* only let unit tests run for 60s */
	alarm(60);
#endif
	mesh_server_t *s = s_server_new(argc, argv);
	mesh_server_run(s);
	mesh_server_destroy(s);

	logger(LOG_INFO, "shutting down");
	log_close();
	return 0;
}
